src/wazuh_db/wdb_upgrade.c

```diff
@@ -86,55 +86,79 @@ wdb_t * wdb_upgrade_global(wdb_t *wdb) {
+         */
         if (wdb_upgrade_check_manager_keepalive(wdb) != 1) {
-            wdb = wdb_backup_global(wdb, -1);
+            if (OS_SUCCESS != wdb_global_create_backup(wdb, output, "-pre_upgrade")) {
+                merror("Creating pre-upgrade Global DB snapshot failed: %s", output);
+                wdb->enabled = false;
+            }
+            else {
+                wdb = wdb_recreate_global(wdb);
+            }
             return wdb;
         }
         break;
     default:
-        if( wdb_metadata_get_entry(wdb, "db_version", db_version) == 1) {
+        if (wdb_metadata_get_entry(wdb, "db_version", db_version) == 1) {
             version = atoi(db_version);
         }
-        else{
+        else {
+            /**
+             * We can't determine if the database should be upgraded. If we allow the usage,
+             * we could have many errors because of an operation over an old database version.
+             * We should block the usage until determine whether we should upgrade or not.
+             */
             mwarn("DB(%s): Error trying to get DB version", wdb->id);
-            wdb = wdb_backup_global(wdb, -1);
+            wdb->enabled = false;
             return wdb;
         }
     }
 
-    char output[OS_MAXSTR + 1] = {0};
-
     if (version < updates_length) {
         if (OS_SUCCESS != wdb_global_create_backup(wdb, output, "-pre_upgrade")) {
-            mwarn("Creating pre-upgrade Global DB snapshot failed: %s", output);
-        }
-
-        bool upgrade_success = TRUE;
-        for (int i = version; i < updates_length; i++) {
-            mdebug2("Updating database '%s' to version %d", wdb->id, i + 1);
-
-            if (wdb_sql_exec(wdb, UPDATES[i]) == -1 || wdb_adjust_global_upgrade(wdb, i)) {
-                mwarn("Failed to update global.db to version %d", i + 1);
-                wdb = wdb_backup_global(wdb, version);
-                upgrade_success = FALSE;
-                break;
-            }
+            merror("Creating pre-upgrade Global DB snapshot failed: %s", output);
+            wdb->enabled = false;
         }
-
-        if (upgrade_success) {
-            if (OS_SUCCESS != wdb_global_create_backup(wdb, output, "-post_upgrade")) {
-                mwarn("Creating post-upgrade Global DB snapshot failed: %s", output);
+        else {
+            for (int i = version; i < updates_length; i++) {
+                mdebug2("Updating database '%s' to version %d", wdb->id, i + 1);
+                if (wdb_sql_exec(wdb, UPDATES[i]) == -1 || wdb_adjust_global_upgrade(wdb, i)) {
+                    char *bkp_name = NULL;
+                    if (OS_INVALID != wdb_global_get_most_recent_backup(&bkp_name) &&
+                        OS_INVALID != wdb_global_restore_backup(&wdb, bkp_name, false, output)) {
+                        merror("Failed to update global.db to version %d. The global.db was restored to the original state.", i + 1);
```

Comment:
As we discussed with the team, this is the only case where we should leave the database enabled.

Reply:
Fixed